File: macos/LiteDesk/Protocol/WireMessages.swift
import Foundation

// Exact wire JSON shapes shared with the Windows native app and the legacy
// Electron app (src/main.js / src/renderer/viewer.js). Field names/casing
// must match byte-for-byte for cross-platform interop.

struct TypeEnvelope: Decodable {
    let type: String
}

struct AuthMessage: Codable {
    let type = "auth"
    let password: String

    enum CodingKeys: String, CodingKey { case type, password }
}

struct AuthOkMessage: Codable {
    let type = "auth-ok"
    let width: Int
    let height: Int

    enum CodingKeys: String, CodingKey { case type, width, height }
}

struct AuthFailMessage: Codable {
    let type = "auth-fail"

    enum CodingKeys: String, CodingKey { case type }
}

struct MouseMoveMessage: Codable {
    let type = "mouse-move"
    let x: Double
    let y: Double

    enum CodingKeys: String, CodingKey { case type, x, y }
}

struct MouseDownMessage: Codable {
    let type = "mouse-down"
    let x: Double
    let y: Double
    let button: String

    enum CodingKeys: String, CodingKey { case type, x, y, button }
}

struct MouseUpMessage: Codable {
    let type = "mouse-up"
    let button: String

    enum CodingKeys: String, CodingKey { case type, button }
}

struct MouseScrollMessage: Codable {
    let type = "mouse-scroll"
    let dx: Double
    let dy: Double

    enum CodingKeys: String, CodingKey { case type, dx, dy }
}

enum MouseMessage {
    case move(x: Double, y: Double)
    case down(x: Double, y: Double, button: String)
    case up(button: String)
    case scroll(dx: Double, dy: Double)

    static func parse(_ data: Data) -> MouseMessage? {
        guard let envelope = try? JSONDecoder().decode(TypeEnvelope.self, from: data) else { return nil }
        let decoder = JSONDecoder()
        switch envelope.type {
        case "mouse-move":
            guard let m = try? decoder.decode(MouseMoveMessage.self, from: data) else { return nil }
            return .move(x: m.x, y: m.y)
        case "mouse-down":
            guard let m = try? decoder.decode(MouseDownMessage.self, from: data) else { return nil }
            return .down(x: m.x, y: m.y, button: m.button)
        case "mouse-up":
            guard let m = try? decoder.decode(MouseUpMessage.self, from: data) else { return nil }
            return .up(button: m.button)
        case "mouse-scroll":
            guard let m = try? decoder.decode(MouseScrollMessage.self, from: data) else { return nil }
            return .scroll(dx: m.dx, dy: m.dy)
        default:
            return nil
        }
    }
}

// Keyboard messages. `code` is a layout-independent physical-key identifier
// using the W3C UIEvents KeyboardEvent.code vocabulary (e.g. "KeyA",
// "ShiftLeft", "ArrowLeft") so a macOS viewer can drive a Windows host and
// vice versa without either side knowing the other's native keycode space —
// see KeyCodeMap.swift for the translation table.
struct KeyDownMessage: Codable {
    let type = "key-down"
    let code: String

    enum CodingKeys: String, CodingKey { case type, code }
}

struct KeyUpMessage: Codable {
    let type = "key-up"
    let code: String

    enum CodingKeys: String, CodingKey { case type, code }
}

enum KeyMessage {
    case down(code: String)
    case up(code: String)

    static func parse(_ data: Data) -> KeyMessage? {
        guard let envelope = try? JSONDecoder().decode(TypeEnvelope.self, from: data) else { return nil }
        let decoder = JSONDecoder()
        switch envelope.type {
        case "key-down":
            guard let m = try? decoder.decode(KeyDownMessage.self, from: data) else { return nil }
            return .down(code: m.code)
        case "key-up":
            guard let m = try? decoder.decode(KeyUpMessage.self, from: data) else { return nil }
            return .up(code: m.code)
        default:
            return nil
        }
    }
}

File: windows/LiteDesk/Protocol/WireMessages.cs
using System.Text.Json.Serialization;

namespace LiteDesk.Protocol;

// Wire-format POCOs. Field names/casing are pinned with [JsonPropertyName] to
// stay byte-for-byte compatible with the existing Electron app's protocol
// (see src/main.js and src/renderer/viewer.js) and with the parallel native
// macOS app, so a Windows host can interoperate with a macOS viewer and vice
// versa over plain ws://.

public abstract class WireMessage
{
    [JsonPropertyName("type")]
    public string Type { get; protected set; } = "";
}

public sealed class AuthMessage : WireMessage
{
    public AuthMessage() { Type = "auth"; }

    [JsonPropertyName("password")]
    public string Password { get; set; } = "";
}

public sealed class AuthOkMessage : WireMessage
{
    public AuthOkMessage() { Type = "auth-ok"; }

    [JsonPropertyName("width")]
    public int Width { get; set; }

    [JsonPropertyName("height")]
    public int Height { get; set; }
}

public sealed class AuthFailMessage : WireMessage
{
    public AuthFailMessage() { Type = "auth-fail"; }
}

public sealed class MouseMoveMessage : WireMessage
{
    public MouseMoveMessage() { Type = "mouse-move"; }

    [JsonPropertyName("x")]
    public double X { get; set; }

    [JsonPropertyName("y")]
    public double Y { get; set; }
}

public sealed class MouseDownMessage : WireMessage
{
    public MouseDownMessage() { Type = "mouse-down"; }

    [JsonPropertyName("x")]
    public double X { get; set; }

    [JsonPropertyName("y")]
    public double Y { get; set; }

    [JsonPropertyName("button")]
    public string Button { get; set; } = "left";
}

public sealed class MouseUpMessage : WireMessage
{
    public MouseUpMessage() { Type = "mouse-up"; }

    [JsonPropertyName("button")]
    public string Button { get; set; } = "left";
}

public sealed class MouseScrollMessage : WireMessage
{
    public MouseScrollMessage() { Type = "mouse-scroll"; }

    [JsonPropertyName("dx")]
    public double Dx { get; set; }

    [JsonPropertyName("dy")]
    public double Dy { get; set; }
}

// `Code` is a layout-independent physical-key identifier using the W3C
// UIEvents KeyboardEvent.code vocabulary (e.g. "KeyA", "ShiftLeft",
// "ArrowLeft") so a Windows viewer can drive a macOS host and vice versa
// without either side knowing the other's native keycode space — see
// Native/KeyCodeMap.cs for the translation table.
public sealed class KeyDownMessage : WireMessage
{
    public KeyDownMessage() { Type = "key-down"; }

    [JsonPropertyName("code")]
    public string Code { get; set; } = "";
}

public sealed class KeyUpMessage : WireMessage
{
    public KeyUpMessage() { Type = "key-up"; }

    [JsonPropertyName("code")]
    public string Code { get; set; } = "";
}

// Bidirectional keepalive/latency probe: either side may send a ping at any
// time; the other side replies with a pong echoing the same `ts` unchanged,
// so the sender can compute round-trip time as (now - ts). `ts` is Unix
// epoch seconds (a double, sub-second precision) — must match the macOS
// app's Date().timeIntervalSince1970 encoding byte-for-byte.
public sealed class PingMessage : WireMessage
{
    public PingMessage() { Type = "ping"; }

    [JsonPropertyName("ts")]
    public double Ts { get; set; }
}

public sealed class PongMessage : WireMessage
{
    public PongMessage() { Type = "pong"; }

    [JsonPropertyName("ts")]
    public double Ts { get; set; }
}
